src/hotspot/os/linux/cgroupSubsystem_linux.cpp

```diff
@@ -558,7 +558,25 @@ jlong CgroupSubsystem::memory_limit_in_bytes() {
   if (!memory_limit->should_check_metric()) {
     return memory_limit->value();
   }
+  jlong phys_mem = os::Linux::physical_memory();
+  log_trace(os, container)("total physical memory: " JLONG_FORMAT, phys_mem);
   jlong mem_limit = read_memory_limit_in_bytes();
+
+  // sanity checking
+  if (mem_limit <= 0 || mem_limit >= phys_mem) {
+    const char *reason;
+    if (mem_limit >= phys_mem) {
+      reason = "ignored";
+    } else if (OSCONTAINER_ERROR == mem_limit) {
+      reason = "failed";
+    } else {
+      reason = "unlimited";
+    }
+    log_debug(os, container)("container memory limit %s: " JLONG_FORMAT ", using host value " JLONG_FORMAT,
+                             reason, mem_limit, phys_mem);
+    mem_limit = mem_limit == -2 ? -2 : -1;
```

Comment:
I take it, the reason for this line is so that we have appropriate info for the log line? A slightly clearer way (to me) would be to use another local, `read_mem_limit` for example, that is used for logging and set `mem_limit` to the desired value in the conditional branches before. Also, I think this code deserves a comment. For example:

```
    jlong read_mem_limit = mem_limit;
    if (mem_limit >= phys_mem) {
      // Exceeding physical memory counts as unlimited. cg v1 is bound
      // above by phys_mem (as there is no 'max' in interface files), but cg v2
      // might return a value > phys_mem if the container engine was started
      // with a memory flag exceeding phys_mem, so we need to account for it
      // here.
      reason = "ignored";
      mem_limit = -1;
    } else if (OSCONTAINER_ERROR == mem_limit) {
      reason = "failed";
    } else {
      assert(mem_limit == -1, "Expected unlimited");
      reason = "unlimited";
    }
    log_debug(os, container)("container memory limit %s: " JLONG_FORMAT ", using host value " JLONG_FORMAT,
                             reason, read_mem_limit, phys_mem);
```

Reply:
> I take it, the reason for this line is so that we have appropriate info for the log line?

Yes, that's why. I'm very happy to eliminate the ternary operator. Addressed in [d13fae1](https://github.com/openjdk/jdk/pull/9880/commits/d13fae1884ac1d48c71e490f0459b7822a599968). Thanks!